Feature local config

Lansoweb opened the pull request:
Provide a narrative description of what you are trying to accomplish:

- [ ] Are you fixing a bug?
  - [ ] Detail how the bug is invoked currently.
  - [ ] Detail the original, incorrect behavior.
  - [ ] Detail the new, expected behavior.
  - [ ] Base your feature on the `master` branch, and submit against that branch.
  - [ ] Add a regression test that demonstrates the bug, and proves the fix.
  - [ ] Add a `CHANGELOG.md` entry for the fix.

- [X] Are you creating a new feature?
  - [X] Why is the new feature needed? What purpose does it serve?
  - [X] How will users use the new feature?
  - [X] Base your feature on the `develop` branch, and submit against that branch.
  - [X] Add only one feature per pull request; split multiple features over multiple pull requests
  - [ ] Add tests for the new feature.
  - [X] Add documentation for the new feature.
  - [X] Add a `CHANGELOG.md` entry for the new feature.

The `config` command provides a config file per project. So you can have a global github token
and a gitlab one for a project.

- [ ] Is this related to quality assurance?
  <!-- Detail why the changes are necessary -->

- [ ] Is this related to documentation?
  <!-- Is it a typographical and/or grammatical fix? -->
  <!-- Is it new documentation? -->

weierophinney commented:
@Lansoweb I took the liberty of finishing this up for you, as I merged #24 earlier today. In particular:

- I added tests for the `Config` and `ConfigCommand` classes.
- I renamed `Provider\GetProviderTrait` to `GetConfigValuesTrait`, updated it to use the `ConfigFileTrait`, and added the methods `prepareConfig()` and `getToken()`. I updated `getProvider()` to also accept a `Config` instance, which is queried for the provider. Essentially, you call `prepareConfig()` to get a `Config` instance, and then pull the token and provider from that; the method grabs the configuration, and then overrides it with values passed via options if present.
- I updated the `ReleaseCommand` and `EntryCommand` to use the methods introduced above for the purpose of getting tokens and/or providers.
- I ensured all tests pass.
- I did a few other minor refactors for readability and maintainability.

Thanks a ton for this!

Lansoweb commented:
> @Lansoweb I took the liberty of finishing this up for you, as I merged #24 earlier today. In particular:
> 
> * I added tests for the `Config` and `ConfigCommand` classes.
> * I renamed `Provider\GetProviderTrait` to `GetConfigValuesTrait`, updated it to use the `ConfigFileTrait`, and added the methods `prepareConfig()` and `getToken()`. I updated `getProvider()` to also accept a `Config` instance, which is queried for the provider. Essentially, you call `prepareConfig()` to get a `Config` instance, and then pull the token and provider from that; the method grabs the configuration, and then overrides it with values passed via options if present.
> * I updated the `ReleaseCommand` and `EntryCommand` to use the methods introduced above for the purpose of getting tokens and/or providers.
> * I ensured all tests pass.
> * I did a few other minor refactors for readability and maintainability.
> 
> Thanks a ton for this!

Thanks a lot for this tool :)
